[env-vars] Make environment variable handling more robust

peter-b opened the pull request:
These are the changes that avoid subtle engine breakage when LiveCode is started with a weird environment array.  It fixes several IDE bugs on my Linux machines/VMs.
- Always use the _first_ definition of each variable in the environment (to match bash)
- Don't import variables that aren't valid LiveCode tokens as `$<name>` variables
- Be more consistent about what's considered a valid `$<name>` variable
- Handle degenerate environment variables that aren't in `name=value` format

These changes **do not** (because I don't have time):
- Synchronise `setenv()` changes back to `MCenvironmentvariables`
- Provide syntax to inspect/modify environment variables that weren't imported
- Ensure that only environment variables visible to LiveCode scripts get passed to subprocesses.

peter-b committed:
[env-vars] Add MCModeHasEnvironmentVariables().

Add a check function for evaluating whether the current mode should
capture environment variables at startup.

peter-b committed:
[env-vars] Add MCU_is_token().

Add a utility function that checks whether a given string is a valid
LiveCode script token.

peter-b committed:
[env-vars] Filter out "bad" environment variables.

Promote environment variables to `$<name>` global variables iff the
`<name>` is a valid LiveCode script token.

peter-b committed:
[env-vars] Capture raw environment variables at startup.

Capture raw environment variables into the `MCenvironmentvariables`
global variable during `X_open()` invocation.

* If multiple environment variables with the same name are found, the
  first is used.

* If an environment variable's whole name is a valid token starting
  with an uppercase letter, it gets turned into an global `$<name>`
  variable.

* If an environment string is found which is not in the `name=value`
  format, the whole variable is treated as a variable name, with an
  empty value.  However, it never becomes a global variable, and has
  lower priority than "proper" environment variables.  For example, if
  the environment contains `PATH` followed by `PATH=/usr/bin`, then
  the `MCenvironmentvariables` array will contain `PATH`:`/usr/bin`.

peter-b committed:
[env-vars] Make sure MCVariable and X_open() agree on env globals

`MCVariable` thinks that global variables from the environment
(`$<name>`) can have any name that doesn't start with a digit or `#`.
`X_open()` thinks the names have to start with a capital letter.

Relax `X_open()`'s rules to match `MCVariable`'s.

peter-b committed:
[env-vars] MCU_is_token(): Clear MCerrorlock.

Bug introduced in commit 39eb28e.

peter-b committed:
[env-vars] MCRange is up to UINDEX_MAX, not UINT32_MAX.

livecodeali commented:
@livecode-vulcan review ok 125c97f

livecode-vulcan commented:
💙 review by @livecodeali ok 125c97f8123889032a005b5b7581da52413f9350

livecode-vulcan commented:
😎 test success 125c97f8123889032a005b5b7581da52413f9350
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/31)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/31)
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/31)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/31)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/33)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/31)

runrevmark commented on engine/src/mode.h:
```diff
@@ -84,6 +84,13 @@ bool MCModeIsExecutableFirstArgument(void);
+//
+// This hook is called by X_open.
+//
+bool MCModeHasEnvironmentVariables(void);
```
What's the reason this is needed? All the implementations return true...

peter-b replied:
I put it in to match the `MCcommandarguments`. Shall I remove it?

runrevmark commented on engine/src/globals.cpp:
```diff
@@ -864,6 +869,174 @@ void X_clear_globals(void)
+	/* Create a global variable for the environment variable, but
+	 * only if the variable name is a valid token beginning with a
+	 * capital letter. */
+	unichar_t t_first = MCStringGetCharAtIndex(p_name_str, 0);
```
Where's the 'isupper' check?

To be fair I'm not entirely clear why that check was there in the first place... Is there something wrong with using env-vars which start with a lower-case letter?

peter-b replied:
Explained in the commit notes for d4e73c2b504f7d5c879896e91ed21f96fb6cf533.

peter-b committed:
[env-vars] Correct some comments to match rules for $<name> variables

Commit d4e73c2 changed the rules for when environment variables are
permitted to be turned into global `$<name>` variables, but missed
updating some comments.

runrevmark commented:
@livecode-vulcan review ok 80bd029

livecode-vulcan commented:
💙 review by @runrevmark ok 80bd02997c14fff9053b6f5497db94582c898bb0

livecode-vulcan commented:
😎 test success 80bd02997c14fff9053b6f5497db94582c898bb0
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/32)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/32)
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/32)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/32)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/34)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/32)
